feature: add floating window footer

echasnovski opened the pull request:
Sometimes it is useful to be able to not only show information in the top border (a.k.a. title), but also in the bottom one (a.k.a. footer).

This PR adds the ability to set footer similar to title. Added functionality:
- New allowed `footer` and `footer_pos` fields for floating window options, which behave similar to `title` and `title_pos` respectively.
- New `FloatFooter` highlight group to go along with `FloatTitle`.

What I actually did:
- Refactored code using `title` and/or `title_pos` into reusable chunks (where it seemed reasonable).
- Updated those code chunks to behave accordingly for footer. Tried to cover any place where `title` is used for floating window, but could have missed some.
- Added tests similar to testing of `title` and `title_pos`.
- Added new `FloatBorder` highlight group updating any its occurrence to also have `FloatFooter`.

Notes/questions:
- I added separate test for showing title and footer together. Not sure if it is needed, but opted on the more test coverage.
- Couldn't find a test for actual usage of `FloatTitle`, so didn't add one for `FloatBorder`.

Here is how it looks:

![neovim_float-win-footer](https://github.com/neovim/neovim/assets/24854248/527abb5e-46ae-4624-97c9-123490c3b2c0)

<details><summary>Code to reproduce screenshot</summary>

```lua
local buf_id = vim.api.nvim_create_buf(false, true)
vim.api.nvim_buf_set_lines(buf_id, 0, -1, false, { 'Very important text' })

vim.api.nvim_set_hl(0, 'FloatFooter', { link = 'DiagnosticFloatingHint' })

local row = 1

local open_win = function(config)
  local ref_config = {
    relative = 'editor',
    anchor = 'NE',
    row = row,
    col = vim.o.columns,
    width = 40,
    height = 1,
    style = 'minimal',
    border = 'single',
    title = 'Old boring title',
    title_pos = 'center',
    footer = 'New shiny footer',
    footer_pos = 'center',
  }

  vim.api.nvim_open_win(buf_id, false, vim.tbl_extend('force', ref_config, config or {}))
  row = row + 4
end

open_win()
open_win({ footer = { { 'New shiny footer' } } })
open_win({
  footer = {
    { 'New ',   'DiagnosticFloatingInfo' },
    { 'shiny ', 'DiagnosticFloatingWarn' },
    { 'footer', 'DiagnosticFloatingError' },
  },
})
open_win({ footer_pos = 'left' })
open_win({ footer_pos = 'right' })
```

</details>

echasnovski commented on src/nvim/drawscreen.c:
```diff
@@ -778,10 +778,17 @@ static void win_redr_border(win_T *wp)
       grid_put_schar(grid, irow + adj[0], i + adj[3], chars[ic], attrs[ic]);
     }
+
+    if (wp->w_float_config.footer) {
```
@bfredl: I put code here solely based on trial/error and some sense of symmetry. It seems to work, but I might miss something..

bfredl replied:
Symmetry with title is correct here:)

gpanders commented on runtime/doc/news.txt:
```diff
@@ -145,6 +145,12 @@ The following new APIs and features were added.
 
 • Improved messages for type errors in `vim.api.*` calls (including `opts` params)
 
+• Floating windows can now show footer with new `footer` and `footer_pos`
+  config fields.
+
+• New highlight group |hl-FloatFooter| to highlight floating window footer
+  separately from title.
```
IMO these don't need to be separate.

```suggestion
• Floating windows can now show a footer with the new `footer` and `footer_pos`
  config fields. |hl-FloatFooter|
```

gpanders commented on src/nvim/api/win_config.c:
```diff
@@ -145,11 +145,18 @@
+///   - title: Title (optional) in window border, string or list.
+///     List should consist of `[text, highlight]` tuples.
+///     If string, the default highlight group is `FloatTitle`.
+///   - title_pos: Title position. Must be set with `title` option.
+///     Value can be one of `"left"` `"center"` `"right"`.
+///     Default is `"left"`.
+///   - footer: Footer (optional) in window border, string or list.
+///     List should consist of `[text, highlight]` tuples.
+///     If string, the default highlight group is `FloatFooter`.
+///   - footer_pos: Footer position. Must be set with `footer` option.
+///     Value can be one of `"left"` `"center"` `"right"`.
+///     Default is `"left"`.
```
```suggestion
///   - title_pos: Title position. Must be set with `title` option.
///     Value can be one of "left", "center", or "right"`.
///     Default is "left".
///   - footer: Footer (optional) in window border, string or list.
///     List should consist of `[text, highlight]` tuples.
///     If string, the default highlight group is `FloatFooter`.
///   - footer_pos: Footer position. Must be set with `footer` option.
///     Value can be one of "left", "center", or "right".
///     Default is "left".
```

gpanders commented on src/nvim/api/win_config.c:
```diff
@@ -240,6 +247,49 @@ void nvim_win_set_config(Window window, Dict(float_config) *config, Error *err)
+  PUT(config, field_name, ARRAY_OBJ(bordertext));
+
+  char *pos;
+  if (align == kAlignLeft) {
+    pos = "left";
+  } else if (align == kAlignCenter) {
+    pos = "center";
+  } else {
+    pos = "right";
+  }
```
Personally I prefer to use `switch` statements for enums. This is (in my opinion) more semantically clear, and also has the benefit that if a new enum variant is ever added, the compiler will warn you if the switch is non-exhaustive.

```suggestion
  switch (align) {
  case kAlignLeft:
    pos = "left";
    break;
  case kAlignCenter:
    pos = "center";
    break;
  case kAlignRight:
    pos = "right";
    break;
  }
```

(I know you didn't write this code, you just moved it, but that's as good a time as any to make minor changes like this).

gpanders commented on src/nvim/api/win_config.c:
```diff
@@ -240,6 +247,49 @@ void nvim_win_set_config(Window window, Dict(float_config) *config, Error *err)
+  AlignTextPos align;
+  char* field_name;
+  char* field_pos_name;
+  if (bordertext_type == kBorderTextTitle) {
+    chunks = fconfig->title_chunks;
+    align = fconfig->title_pos;
+    field_name = "title";
+    field_pos_name = "title_pos";
+  } else if (bordertext_type == kBorderTextFooter) {
+    chunks = fconfig->footer_chunks;
+    align = fconfig->footer_pos;
+    field_name = "footer";
+    field_pos_name = "footer_pos";
+  }
```
These two `if` statements cover all variants of `BorderTextType`, but if a new variant is ever added and used you will end up with a read of an uninitialized variable (`field_name` and `field_pos_name`).

Suggestion: either (1) use a `switch` statement (so the compiler will warn you if a new unhandled variant is ever added), or (2) initialize `field_name` and `field_pos_name` to `NULL`, and below this `if` statement add `assert(field_name != NULL)`.

gpanders commented on src/nvim/api/win_config.c:
```diff
@@ -240,6 +247,49 @@ void nvim_win_set_config(Window window, Dict(float_config) *config, Error *err)
+{
+  VirtText chunks;
+  AlignTextPos align;
+  char* field_name;
+  char* field_pos_name;
```
```suggestion
  char *field_name;
  char *field_pos_name;
```

gpanders commented on src/nvim/api/win_config.c:
```diff
@@ -381,54 +415,80 @@ static bool parse_float_bufpos(Array bufpos, lpos_T *out)
+  VirtText *chunks;
+  int *width;
+  int default_hl_id;
+  if (bordertext_type == kBorderTextTitle) {
+    is_present = &fconfig->title;
+    chunks = &fconfig->title_chunks;
+    width = &fconfig->title_width;
+    default_hl_id = syn_check_group(S_LEN("FloatTitle"));
+  } else if (bordertext_type == kBorderTextFooter) {
+    is_present = &fconfig->footer;
+    chunks = &fconfig->footer_chunks;
+    width = &fconfig->footer_width;
+    default_hl_id = syn_check_group(S_LEN("FloatFooter"));
+  }
```
Similar to my other comment, if a new `BorderTextType` variant is ever added you could end up with an uninitialized read from this `if` branch. Same advice as before: either convert to `switch` or initialize pointers to `NULL` and `assert` after the if statement.

gpanders commented on src/nvim/api/win_config.c:
```diff
@@ -381,54 +415,80 @@ static bool parse_float_bufpos(Array bufpos, lpos_T *out)
 {
-  if (title_pos.size == 0) {
-    fconfig->title_pos = kAlignLeft;
+  AlignTextPos* align;
+  if (bordertext_type == kBorderTextTitle) {
+    align = &fconfig->title_pos;
+  } else if (bordertext_type == kBorderTextFooter) {
+    align = &fconfig->footer_pos;
+  }
```
Same

echasnovski commented:
@gpanders, @ii14: I've added all your review suggestions in "to be later rebased" commits. Thanks for taking time to review!

echasnovski committed:
refactor(float): extract "title" and "title_pos" handling

echasnovski committed:
feat(float): implement footer

Problem: Now way to show text at the bottom part of floating window
  border (a.k.a. "footer").

Solution: Allows `footer` and `footer_pos` config fields similar to
  `title` and `title_pos`.

echasnovski committed:
feat(highlight): add `FloatFooter` highlight group

Problem: No clear separation of floating title and footer highlighting.

Solution: Add new `FloatFooter` highlight group.

bfredl commented on src/nvim/drawscreen.c:
```diff
@@ -699,20 +699,32 @@ void end_search_hl(void)
   }
 }
 
+int win_border_get_text_col(int total_col, int text_width, AlignTextPos align)
```
maybe `win_get_bordertext_col` as "bordertext" seems to be used regularly for either header/footer.

echasnovski replied:
I opted for not using `bordertext` as it doesn't depend on `bordertext_type`.

bfredl replied:
But it still is only used for these two things. I don't think there is anything more it could be used for as we now have filled both the top and bottom border parts with (potential) text, right?

echasnovski replied:
> But it still is only used for these two things. I don't think there is anything more it could be used for as we now have filled both the top and bottom border parts with (potential) text, right?

I had a crazy idea of maybe allowing text on left and right parts, but can't see useful use case.

glepnir commented:
can use a more generic parameter instead of adding new ones ?  At the time I rarely saw the use of the bottom titile so I didn't implement it

echasnovski commented:
> can use a more generic parameter instead of adding new ones ? At the time I rarely saw the use of the bottom title so I didn't implement it

I don't think any more text is reasonable to add to left and right sides.
Having two separate fields allows setting both title and footer.

glepnir commented:
something like `title_side = { 'top', 'foot' }`  or `ttile_side = 'top' or title_side = 'foot'`   left and right don't matter. just avoid increase parameters.  more generic

echasnovski commented:
> something like `title_side = { 'top', 'foot' }` or `ttile_side = 'top' or title_side = 'foot'` left and right don't matter. just avoid increase parameters. more generic

It would still require two parameters for title and footer texts while breaking backwards compatibility. I think introducing two new dedicated parameters is the most straightforward approach here.

max397574 commented:
> something like `title_side = { 'top', 'foot' }` or `titile_side = 'top' or title_side = 'foot'` also `VirtText` struct left and right don't matter. just avoid increase parameters. more generic
> 
> ```lua
>  title_side = { 'top', 'foot' },
>   title = { top = { 'foo', 'hl'}, foot = {'bar', 'hl} }
>   title_pos = { top = 'center', foot= 'left' }
> ```

that also makes things much more complicated when you just want one of those

glepnir commented:
Why would it break compatibility. You can check it is array or dictionary.   Just my opinion if bfredl is ok with this implementation.

glepnir commented:
> that also makes things much more complicated when you just want one of those

It doesn't. compatibility can also be maintained. but need some codes on type check and convert .

max397574 commented:
don't like the idea of having two so completely different syntaxes
and even if you still can use the old one I guess the new one will be the one that's intended to be used
and `title_pos={top="left"}` is just ugly in my opinion

also I don't really like having footer under a "title" field but perhaps the native English speakers here have a different opinion on this

glepnir commented:
title_pos also can use index.  1 is top 2 is foot.  Increase the number of parameters horizontally. Never in a good way. Maybe in the future someone wants to add left or right. Then continue to increase the parameters ？  No one knows what will happen in the future. Like I thought no one would use foot so I didn't add foot. Because I don't like it . But some people like it. So here is the pr of foot :)

bfredl commented on test/functional/ui/float_spec.lua:
```diff
@@ -2033,6 +2066,368 @@ describe('float window', function()
+          {0:~                                       }|
+        ## grid 3
+                                                  |
+        ## grid 5
```
note: as #24824 was merged, some adjustments are needed here. They should be quite regular tho. In these added tests `grid 5` should be changed into `grid 4`, the array index `[5]` into `[4]` and finally `id = 1002` into `id = 1001` .

echasnovski replied:
Thanks for the heads up. Done.

echasnovski commented:
I have removed addition of `FloatFooter` to 'highlight_defs.h' because it caused halting 'ui/embed_spec.lua' functional test on Windows. This has nothing to do with this PR's functionality per se, see CI results of #24888.
